File: php/src/LombokAlgoritma/Graph/TopologicalSort.php
<?php

// LombokAlgoritma — topological sort, Kahn (SPEC §9.7)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

declare(strict_types=1);

namespace LombokAlgoritma\Graph;

/** Kahn's algorithm with a FIFO queue. O(V + E). */
final class TopologicalSort
{
    /**
     * Queue seeded with the in-degree-0 nodes ascending; neighbours released in edge-list order.
     * Returns `[]` when the graph has a cycle.
     *
     * @return list<int>
     */
    public static function sort(Graph $g): array
    {
        $adj = $g->adjacency();
        $indegree = $g->nodes > 0 ? array_fill(0, $g->nodes, 0) : [];
        foreach ($g->edges as $e) {
            $indegree[$e->to]++;
        }
        $queue = [];
        for ($i = 0; $i < $g->nodes; $i++) {
            if ($indegree[$i] === 0) {
                $queue[] = $i;
            }
        }
        for ($head = 0; $head < count($queue); $head++) {
            foreach ($adj[$queue[$head]] as [$to]) {
                $indegree[$to]--;
                if ($indegree[$to] === 0) {
                    $queue[] = $to;
                }
            }
        }
        return count($queue) === $g->nodes ? $queue : [];
    }
}

File: rust/lombokalgoritma/src/graph/topological_sort.rs
// LombokAlgoritma — topological sort, Kahn (SPEC §9.7)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
use super::types::{adjacency, Graph};
use crate::Result;
use alloc::vec;
use alloc::vec::Vec;

/// Kahn's algorithm with a FIFO queue seeded with the in-degree-0 nodes ascending; neighbours
/// are released in edge-list order. Returns an empty vector when the graph has a cycle. O(V + E).
///
/// # Errors
/// [`crate::Error::OutOfRange`] for an edge endpoint outside `[0, n)`.
pub fn topological_sort(g: &Graph) -> Result<Vec<usize>> {
    let adj = adjacency(g)?;
    let mut indeg = vec![0usize; g.nodes];
    for e in &g.edges {
        indeg[e.to] += 1;
    }
    let mut queue: Vec<usize> = (0..g.nodes).filter(|&i| indeg[i] == 0).collect();
    let mut head = 0;
    while head < queue.len() {
        let u = queue[head];
        head += 1;
        for &(to, _) in &adj[u] {
            indeg[to] -= 1;
            if indeg[to] == 0 {
                queue.push(to);
            }
        }
    }
    if queue.len() == g.nodes {
        Ok(queue)
    } else {
        Ok(Vec::new())
    }
}
